content/posts/redis-modules-into-arm-land-part1.md

```diff
@@ -0,0 +1,275 @@
+---
+title: "Getting Redis Modules into ARM land - Part 1"
+date: 2019-11-13
+authors:
+  - Rafi Einstein
+tags: ["redisedge", "arm"]
+---
+
+Our motivation at Redis Labs of getting into ARM land and dragging Redis Modules along was RedisEdge.
+Redis, of course, has long been native in this land, in both glibc and alpine/musl variants.
+Redis Modules, although already on the multi-platform scene (running on various Linux distributions and supporting macOS, mainly for the sake of development experience), have been more of an enterprise/data-center thing, all until RedisEdge came along, targeting IoT devices.
+In this series of posts, I'll describe our vision of ARM platform support and the developer user experience, as well as the steps we took along the path, as the way was as valuable as the outcome.
+
+If you'll follow through, you'll end up with a fully functional ARM build laboratory.
+
+## Inside RedisEdge
+
+![redisedge1](/redis-modules-arm-1.png)
+
+But first, let's take a look at RedisEdge.
+RedisEdge is not a Redis module, but an aggregate of three Redis modules: RedisGears, RedisAI, and RedisTimeSeries.
+It is distributed as a Docker image, which is based on Redis Server 5.0. Thus, one can simply pull the image, run it, and start issuing Redis commands, load models into RedisAI, and execute Python gears scripts on RedisGears.
+Although one can easily get Docker out of the equation (by installing a Redis server and copying Redis Modules files), we'll see that Docker does actually provide significant added value, and it is worthwhile keeping it around.
+
+![redis-edge-1](/redis-modules-arm-2.png)
+
+*Inside RedisEdge: modules structure*
+
+Let's now take a look at each of the components of RedisEdge, to try to figure out what would it take to have them ported to ARM.
+First, **[RedisTimeSeries](https://github.com/RedisTimeSeries/RedisTimeSeries)**. 
+It's a simple C library, built with make. Not even a configure script. No problems there.
+Next is [**RedisGears**](https://github.com/RedisGears/RedisGears). It's too a C library built with plain make, but it does use an embedded Python 3.7 interpreter, which is built from source. This requires running automake to generate a platform-specific makefile.
+Finally, [**RedisAI**](https://github.com/RedisAI/RedisAI). It's a C library built with CMake, and it includes modular "engines" that allow abstraction and encapsulation of AI libraries like [TensorFlow](https://github.com/tensorflow/tensorflow), [PyTorch](https://github.com/pytorch/pytorch), and [ONNXRuntime](https://github.com/microsoft/onnxruntime), in their C library form (most users typically use them in Python), with PyTorch and ONNXRuntime not officially supporting ARM.

```

Comment:
Change to "It's a C library built with CMake, and it includes modular "engines" that allow abstraction and encapsulation of AI libraries like [TensorFlow](https://github.com/tensorflow/tensorflow), [PyTorch](https://github.com/pytorch/pytorch), and [ONNXRuntime](https://github.com/microsoft/onnxruntime) in their C library form—most users typically use them in Python, with PyTorch and ONNXRuntime not officially supporting ARM.